File: ta4j-core/src/test/java/org/ta4j/core/criteria/EnterAndHoldReturnCriterionTest.java
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2017-2024 Ta4j Organization & respective
 * authors (see AUTHORS)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
package org.ta4j.core.criteria;

import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;
import static org.ta4j.core.TestUtils.assertNumEquals;

import java.util.List;

import org.junit.Test;
import org.ta4j.core.Position;
import org.ta4j.core.Trade;
import org.ta4j.core.Trade.TradeType;
import org.ta4j.core.backtest.BackTestTradingRecord;
import org.ta4j.core.mocks.MockBarSeriesBuilder;
import org.ta4j.core.num.NumFactory;

public class EnterAndHoldReturnCriterionTest extends AbstractCriterionTest {

  public EnterAndHoldReturnCriterionTest(final NumFactory numFactory) {
    super(
        params -> params.length == 0
                  ? EnterAndHoldReturnCriterion.buy()
                  : new EnterAndHoldReturnCriterion((TradeType) params[0]),
        numFactory
    );
  }


  @Test
  public void calculateWithEmpty() {
    final var series = new MockBarSeriesBuilder().withNumFactory(this.numFactory).withData(List.of()).build();
    final var tradingRecord = new BackTestTradingRecord();
    final var buyAndHold = getCriterion();
    final var sellAndHold = getCriterion(TradeType.SELL);

    final var buyAndHoldResult = buyAndHold.calculate(series, tradingRecord);
    final var sellAndHoldResult = sellAndHold.calculate(series, tradingRecord);

    assertNumEquals(1, buyAndHoldResult);
    assertNumEquals(1, sellAndHoldResult);

  }


  @Test
  public void calculateOnlyWithGainPositions() {
    final var series = new MockBarSeriesBuilder().withNumFactory(this.numFactory)
        .withData(100, 105, 110, 100, 95, 105)
        .build();
    final var tradingRecord = new BackTestTradingRecord(Trade.buyAt(0, series), Trade.sellAt(2, series),
        Trade.buyAt(3, series), Trade.sellAt(5, series)
    );

    final var buyAndHold = getCriterion();
    assertNumEquals(1.05, buyAndHold.calculate(series, tradingRecord));

    final var sellAndHold = getCriterion(TradeType.SELL);
    assertNumEquals(0.95, sellAndHold.calculate(series, tradingRecord));
  }


  @Test
  public void calculateOnlyWithLossPositions() {
    final var series =
        new MockBarSeriesBuilder().withNumFactory(this.numFactory).withData(100, 95, 100, 80, 85, 70).build();
    final var tradingRecord = new BackTestTradingRecord(
        Trade.buyAt(0, series),
        Trade.sellAt(1, series),
        Trade.buyAt(2, series),
        Trade.sellAt(5, series)
    );

    final var buyAndHold = getCriterion();
    assertNumEquals(0.7, buyAndHold.calculate(series, tradingRecord));

    final var sellAndHold = getCriterion(TradeType.SELL);
    assertNumEquals(1.3, sellAndHold.calculate(series, tradingRecord));
  }


  @Test
  public void calculateWithNoPositions() {
    final var series =
        new MockBarSeriesBuilder().withNumFactory(this.numFactory).withData(100, 95, 100, 80, 85, 70).build();

    final var buyAndHold = getCriterion();
    assertNumEquals(1.0, buyAndHold.calculate(series, new BackTestTradingRecord()));

    final var sellAndHold = getCriterion(TradeType.SELL);
    assertNumEquals(1.0, sellAndHold.calculate(series, new BackTestTradingRecord()));
  }


  @Test
  public void calculateWithOnePositions() {
    final var series = new MockBarSeriesBuilder().withNumFactory(this.numFactory).withData(100, 105).build();
    final var position = new Position(Trade.buyAt(0, series), Trade.sellAt(1, series));
    final var buyAndHold = getCriterion();
    assertNumEquals(105d / 100, buyAndHold.calculate(series, position));

    final var sellAndHold = getCriterion(TradeType.SELL);
    assertNumEquals(0.95, sellAndHold.calculate(series, position));
  }


  @Test
  public void betterThan() {
    final var buyAndHold = getCriterion();
    assertTrue(buyAndHold.betterThan(numOf(1.3), numOf(1.1)));
    assertFalse(buyAndHold.betterThan(numOf(0.6), numOf(0.9)));

    final var sellAndHold = getCriterion(TradeType.SELL);
    assertTrue(sellAndHold.betterThan(numOf(1.3), numOf(1.1)));
    assertFalse(sellAndHold.betterThan(numOf(0.6), numOf(0.9)));
  }
}

File: ta4j-core/src/test/kotlin/org/ta4j/core/criteria/EnterAndHoldReturnCriterionTest.kt
/**
 * The MIT License (MIT)
 *
 * Copyright (c) 2017-2022 Ta4j Organization & respective
 * authors (see AUTHORS)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
package org.ta4j.core.criteria

import org.junit.Assert
import org.junit.Test
import org.ta4j.core.*
import org.ta4j.core.Trade.Companion.buyAt
import org.ta4j.core.Trade.Companion.sellAt
import org.ta4j.core.Trade.TradeType
import org.ta4j.core.mocks.MockBarSeries
import org.ta4j.core.num.Num
import java.util.function.Function

class EnterAndHoldReturnCriterionTest(numFunction: Function<Number?, Num>) : AbstractCriterionTest(
    CriterionFactory { params: Array<out Any?> ->
        if (params.size == 0) EnterAndHoldReturnCriterion() else EnterAndHoldReturnCriterion(
            (params[0] as TradeType?)!!
        )
    }, numFunction
) {
    @Test
    fun calculateOnlyWithGainPositions() {
        val series = MockBarSeries(numFunction, 100.0, 105.0, 110.0, 100.0, 95.0, 105.0)
        val tradingRecord: TradingRecord = BaseTradingRecord(
            buyAt(0, series), sellAt(2, series),
            buyAt(3, series), sellAt(5, series)
        )
        val buyAndHold = getCriterion()
        TestUtils.assertNumEquals(1.05, buyAndHold!!.calculate(series, tradingRecord))
        val sellAndHold = getCriterion(TradeType.SELL)
        TestUtils.assertNumEquals(0.95, sellAndHold!!.calculate(series, tradingRecord))
    }

    @Test
    fun calculateOnlyWithLossPositions() {
        val series = MockBarSeries(numFunction, 100.0, 95.0, 100.0, 80.0, 85.0, 70.0)
        val tradingRecord: TradingRecord = BaseTradingRecord(
            buyAt(0, series), sellAt(1, series),
            buyAt(2, series), sellAt(5, series)
        )
        val buyAndHold = getCriterion()
        TestUtils.assertNumEquals(0.7, buyAndHold!!.calculate(series, tradingRecord))
        val sellAndHold = getCriterion(TradeType.SELL)
        TestUtils.assertNumEquals(1.3, sellAndHold!!.calculate(series, tradingRecord))
    }

    @Test
    fun calculateWithNoPositions() {
        val series = MockBarSeries(numFunction, 100.0, 95.0, 100.0, 80.0, 85.0, 70.0)
        val buyAndHold = getCriterion()
        TestUtils.assertNumEquals(0.7, buyAndHold!!.calculate(series, BaseTradingRecord()))
        val sellAndHold = getCriterion(TradeType.SELL)
        TestUtils.assertNumEquals(1.3, sellAndHold!!.calculate(series, BaseTradingRecord()))
    }

    @Test
    fun calculateWithOnePositions() {
        val series = MockBarSeries(numFunction, 100.0, 105.0)
        val position = Position(buyAt(0, series), sellAt(1, series))
        val buyAndHold = getCriterion()
        TestUtils.assertNumEquals(105.0 / 100, buyAndHold!!.calculate(series, position))
        val sellAndHold = getCriterion(TradeType.SELL)
        TestUtils.assertNumEquals(0.95, sellAndHold!!.calculate(series, position))
    }

    @Test
    fun betterThan() {
        val buyAndHold = getCriterion()
        Assert.assertTrue(buyAndHold!!.betterThan(numOf(1.3)!!, numOf(1.1)!!))
        Assert.assertFalse(buyAndHold.betterThan(numOf(0.6)!!, numOf(0.9)!!))
        val sellAndHold = getCriterion(TradeType.SELL)
        Assert.assertTrue(sellAndHold!!.betterThan(numOf(1.3)!!, numOf(1.1)!!))
        Assert.assertFalse(sellAndHold.betterThan(numOf(0.6)!!, numOf(0.9)!!))
    }
}